### DISCUSSION TITLE
Waveshare ESP32-S3-PICO with Waveshare ResTouch-PICO-LCD-3.5 fails DC/CS out of range

### DISCUSSION THREAD

**Discussion by illigtr** [Bodmer/TFT_eSPI#2842] [New processor/display support or capability request] [answered]
Description of the problem.

Waveshare makes a 3.5 LCD with touchscreen called the ResTouch-PICO-LCD-3.5. This HAT supports the 40 pin Pi Pico, Pi Pico W, Waveshare dev kits: ESP32-S2-Pico and WS ESP32-S3-Pico boards. I can successfully use TFT_eSPI with all the combinations except the S3 due to the limitation of the LCD DS and CS requiring a GPIO of 31 or less. I have even tried to breadboard the S3 to the LCD and change DC and CS to a lower GPIO. Still fails. No screen initialization at all. If any user actaully want to use the S3/LCD combination they are forced to use the pin/GPIO listed below.

IDE
    Arduino v2.2.1

TFT_eSPI library version
    v2.5.0

Board package version
   Espresif v2.0.12

Processor
    Waveshare ESP32-S3-PICO

TFT driver
    Waveshare ResTouch-PICO-LCD-3.5 using ILI9488 and XP2066
    
Interface
    SPI

Connections:

Pin    LCD-3.5            PiPico    S2-PICO    S3-PICO     Notes
11     LCD DC           GPIO8     GPIO16     GPIO33     
12     LCD CS           GPIO9     GPIO17     GPIO34     
14     LCD SPI CLK   GPIO10    GPIO10     GPIO35     
15     LCD SPI MOSI  GPIO11    GPIO11     GPIO36     
16     LCD SPI MISO  GPIO12    GPIO12     GPIO37     
17     LCD BL           GPIO13    GPIO13     GPIO38     
20     LCD RST         GPIO15    GPIO35     GPIO40     
21     TP CS              GPIO16    GPIO36      GPIO42     
22     TP IRQ           GPIO17    GPIO37     GPIO41     not used

TFT_eSPI Test          works         works          fails     

For each configuration I have created a custom user setup.h file.

I know that GPIO 33/34 can be used for DC/CS, as I have compiled a Home Assistant sensor with ESPHome using
the WS ESP32-S3-PICO and ResTouch-PICO-LCD-3.5 with success.

**Comment by illigtr** [comment]
![image](https://github.com/Bodmer/TFT_eSPI/assets/120117710/f571e138-a1ab-4e6e-9c3f-892c61b2c28b)

**Comment by Bodmer** [comment]
These mappings of the SPI pins are supported:
```
                    // Typical board default pins - change to match your board
#define TFT_CS   34 //     10 or 34 (FSPI CS0) 
#define TFT_MOSI 35 //     11 or 35 (FSPI D)
#define TFT_SCLK 36 //     12 or 36 (FSPI CLK)
#define TFT_MISO 37 //     13 or 37 (FSPI Q)

// FSPI port (SPI2) used unless following defined
// #define USE_HSPI_PORT
```
GPIO above 31 can be used for DC, CS and RST

This means the board should work if you swap the SCLK and MOSI connections.

**Comment by illigtr** [comment]
Thank you for confirming that there is no restriction on which GPIO pin can be used for DC, CS and RST. As mentionned, I have tested the S2 with success, but the S3 eludes success. I have tested with and without HSPI. Here is my custom setup.h.

I have tested loading the Read_User_Setup.ino to ensure the settings are read and validated via Serial. However, I only get a blank screen whenever I run any basic screen test, like UTFT_Demo_480x320.ino.


```
// custom setup file using the Waveshare ESP32-S3-PICO board with Pico-ResTouch-LCD 3.5 
//  Updated: Sept 8, 2023

#define ILI9488_DRIVER
#define RPI_DISPLAY_TYPE
#define TFT_DC   33
#define TFT_CS   34
#define TFT_SCLK 35
#define TFT_MOSI 36
#define TFT_MISO 37
#define TFT_BL   38
#define TFT_RST  40
#define TOUCH_CS 42

#define TFT_BACKLIGHT_ON HIGH
#define TFT_INVERSION_ON

#define LOAD_GLCD  
#define LOAD_FONT2 
#define LOAD_FONT4 
#define LOAD_FONT6 
#define LOAD_FONT7 
#define LOAD_FONT8 
#define LOAD_GFXFF 
#define SMOOTH_FONT

//#define USE_HSPI_PORT
#define SPI_TOUCH_FREQUENCY 2500000 // support for XPT2046
```

**Reply by Bodmer** [reply]
The SPI native pins are supported. So as explained above the designers appear to have accidentally swapped SCLK and MOSI... use and wire:

#define TFT_MOSI 35 //     11 or 35 (FSPI D)
#define TFT_SCLK 36 //     12 or 36 (FSPI CLK)

This will not work:
#define TFT_SCLK 35
#define TFT_MOSI 36

**Comment by illigtr** [comment]
I have already attempted your suggestion and the result is still a blank screen. What could I use to further debug this issue?

Here is a photo where I have physically rewired CLK/MOSI, orange and yellow jumpers crossed over, per your instructions so that I force the LCD MOSI to GP35 and CLK to GP36 on the DEV board. Updated my custom setup.h per your suggestion.


![image](https://github.com/Bodmer/TFT_eSPI/assets/120117710/0499eed3-2e4b-4d5a-9b2c-efe257741fb3)

And here is the results of the diagnostic sketch:

```
[code]
TFT_eSPI ver = 2.5.0
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 9488
Display width  = 320
Display height = 480

MOSI    = GPIO 35
MISO    = GPIO 37
SCK     = GPIO 36
TFT_CS   = GPIO 34
TFT_DC   = GPIO 33
TFT_RST  = GPIO 40

TFT_BL           = GPIO 38
TFT_BACKLIGHT_ON = HIGH

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 20.00
[/code]

```
Here is the hardcoded pinout of the ResTouchLCD:

![Pico-ResTouch-LCD-3 5-details-inter](https://github.com/Bodmer/TFT_eSPI/assets/120117710/d6d256fd-ecf5-43bd-beac-dd916e3369a3)

Here is the Waveshare ESP32-S3-PICO pinout:

![ESP32-S3-Pico03](https://github.com/Bodmer/TFT_eSPI/assets/120117710/65e00920-985d-4779-974d-9423c474d300)

If I use the DEV board directly pin-for-pin on the LCD board, I am forced to use GP35 for CLK and CP36 for MOSI. This is where I initially started with the original setup.h sent earlier.

--> CURIOUSLY....under IDE ESPhome  the display does work with these hardcoded pin assignments! 

Here is my ESPhome configuration (shortened for brevity) where the display works and I am using CLK on 35 and MOSI on 36 due to the HAT pin assignments.

```
esphome:
  name: hvac2
  friendly_name: HVAC2

esp32:
  board: esp32-s3-devkitc-1
  framework:
    type: arduino

    
spi:
- id: spi_0
  clk_pin: GPIO18
  mosi_pin: GPIO19
  miso_pin: GPIO16

- id: spi_1
  clk_pin: GPIO35
  mosi_pin: GPIO36
  miso_pin: GPIO37    


display:
  - platform: ili9xxx
    model: ili9488
    spi_id: spi_1
    dc_pin: GPIO33
    cs_pin: GPIO34
    reset_pin: GPIO40
    dimensions: 320X480
    #data_rate: 60Mhz
    update_interval: 1s
    auto_clear_enabled: true
    rotation: 90
    id: ws_tft
    lambda: |-
      it.fill(my_black);
      it.print(0,0,font2, my_white,"my_white");


```

So, in conclusion I have attempted nearly all combinations that I can think of. I have tried:

1.  to modify the pins_arduino.h for the S3.
2.  rewired via jumpers to an alternate  spi configuration (pins_arduino for S3 lists: SS 10, MOSI 11, MISO 13, SCK 12) and modified my setup.h to match
3. reduced SPI frequency
4. attempted HSPI instead of FSPI using three different SPI pin setup

Willing to provide any additional information that may prove useful.

**Comment by illigtr** [comment, marked as answer]
I have done considerable testing and research into this issue. I noticed that there was a fix  (#2725) in the github but that didn't show up in my local Arduino library. I thought I had the most recent version (v2.5.0), but the docs for the fix showed v2.5.30... I pulled a full clone of the library from Github locally and VOILA the Waveshare ResTouch-PICO-3.5 now functions with the S3!

I am not sure if it is specific to the fix of 2725, but is it definitely whatever changed between lib version 2.5.0 and 2.5.31 that I downloaded. You can consider this issue closed.

**Comment by illigtr** [comment]
Just a final note, I tested the WS ESP32-S3-PICO piggybacked directly to the WS ResTouch-PICO-LCD-35 to retest the MOSI/CLK pin configuration again. It works, so it appears that the SPI pins are fully configurable even though the WS designers swapped MOSI/CLK pins.

Here is my working setup.h

```
// custom setup file using the Waveshare ESP32-S3-PICO board with Pico-ResTouch-LCD 3.5 
// Updated: Sept 13, 2023
// ensure TFT_eSPI library is version 2.5.31 or better
// this tested with the cpu board mounted pin-for-pin to the LCD HAT.

#define ILI9488_DRIVER
#define RPI_DISPLAY_TYPE
#define TFT_DC   33 
#define TFT_CS   34 
#define TFT_SCLK 35  // although this doesn't follow standard S3 pinout, it does work
#define TFT_MOSI 36  // although this doesn't follow standard S3 pinout, it does work
#define TFT_MISO 37
#define TFT_BL   38 
#define TFT_RST  40 
#define TOUCH_CS 42

#define TFT_BACKLIGHT_ON HIGH
#define TFT_INVERSION_ON

#define LOAD_GLCD  
#define LOAD_FONT2 
#define LOAD_FONT4 
#define LOAD_FONT6 
#define LOAD_FONT7 
#define LOAD_FONT8 
#define LOAD_GFXFF 
#define SMOOTH_FONT

#define SPI_FREQUENCY   20000000 // do not exceed 20Mhz, or display won't work
#define SPI_TOUCH_FREQUENCY 2500000


```


I ran the full UTFTdemo as well as the diagnostics. Here is the results from the diagnostics.

```

[code]
TFT_eSPI ver = 2.5.31
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = SPI
Display driver = 9488
Display width  = 320
Display height = 480

MOSI    = GPIO 36
MISO    = GPIO 37
SCK     = GPIO 35
TFT_CS   = GPIO 34
TFT_DC   = GPIO 33
TFT_RST  = GPIO 40
TOUCH_CS = GPIO 42

TFT_BL           = GPIO 38
TFT_BACKLIGHT_ON = HIGH

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled

Display SPI frequency = 20.00
Touch SPI frequency   = 2.50
[/code]
```
